File: packages/node/src/compile/todos.ts
/**
 * To-do lists (spec/85 *To-do lists*, spec/20 *t1/todos.json*): a `type: todo`
 * doc's checklist lines, compiled into an artifact so the overview can count
 * them and a search hit can say "2 open" — the engine indexes and never edits.
 * Twin of packages/python/src/brainpick/compile/todos.py.
 */
import { readFileSync } from "node:fs";
import { join } from "node:path";

import { isTodo, type Document } from "../core/bundle";
import { cmpStr } from "../core/canonical";

export { isTodo };

export interface TodoItem {
  done: string | null;
  line: number;
  status: "open" | "done";
  text: string;
}

export interface TodoRecord extends TodoItem {
  path: string;
}

export interface TodosArtifact {
  todos: TodoRecord[];
}

export interface TodoCounts {
  open: number;
  done: number;
}

const ITEM = /^\s*[-*+] \[([ xX])\] (.*)$/;
const DONE_SUFFIX = /\s*\(done:\s*(\d{4}-\d{2}-\d{2})\)\s*$/;
const FENCE = /^\s*(```|~~~)/;

/** The checklist items of a document's full text, in order, with 1-based file
 * lines. `timestamp` (the doc's) dates a done item that carries no
 * `(done: YYYY-MM-DD)` suffix of its own; an open item is never dated. */
export function parseTodoItems(text: string, timestamp: string | null): TodoItem[] {
  const docDate = timestamp ? timestamp.slice(0, 10) : null;
  const items: TodoItem[] = [];
  let inFence = false;
  const lines = text.split(/\r?\n/);
  if (lines.length && lines[lines.length - 1] === "") lines.pop();
  lines.forEach((line, i) => {
    if (FENCE.test(line)) {
      inFence = !inFence;
      return;
    }
    if (inFence) return;
    const match = ITEM.exec(line);
    if (!match) return;
    const status: "open" | "done" = match[1] === " " ? "open" : "done";
    let body = match[2]!.trim();
    let done: string | null = null;
    const suffix = DONE_SUFFIX.exec(body);
    if (suffix) {
      body = body.slice(0, suffix.index).trimEnd();
      done = suffix[1]!;
    }
    if (status === "done" && done === null) done = docDate;
    items.push({ done: status === "done" ? done : null, line: i + 1, status, text: body });
  });
  return items;
}

/** The t1/todos.json payload: every item of every to-do list, sorted by (path, line). */
export function buildTodos(docs: readonly Document[], root: string): TodosArtifact {
  const todos: TodoRecord[] = [];
  for (const doc of [...docs].sort((a, b) => cmpStr(a.path, b.path))) {
    if (!doc.todo) continue;
    const text = readFileSync(join(root, doc.path), "utf8");
    for (const item of parseTodoItems(text, doc.timestamp)) {
      todos.push({ done: item.done, line: item.line, path: doc.path, status: item.status, text: item.text });
    }
  }
  return { todos };
}

/** `{open, done}` over the items — of one list when `path` is given. */
export function todoCounts(todos: readonly TodoRecord[], path?: string): TodoCounts {
  const scoped = path === undefined ? todos : todos.filter((t) => t.path === path);
  return {
    open: scoped.filter((t) => t.status === "open").length,
    done: scoped.filter((t) => t.status === "done").length,
  };
}

File: packages/python/src/brainpick/compile/todos.py
"""To-do lists (spec/85 *To-do lists*, spec/20 *t1/todos.json*): a `type: todo`
doc's checklist lines, compiled into an artifact so the overview can count
them and a search hit can say "2 open" — the engine indexes and never edits."""
from __future__ import annotations

import re
from pathlib import Path

from brainpick.core.bundle import Document, is_todo

__all__ = ["build_todos", "is_todo", "parse_todo_items", "todo_counts"]

_ITEM = re.compile(r"^\s*[-*+] \[([ xX])\] (.*)$")
_DONE_SUFFIX = re.compile(r"\s*\(done:\s*(\d{4}-\d{2}-\d{2})\)\s*$")
_FENCE = re.compile(r"^\s*(```|~~~)")


def parse_todo_items(text: str, timestamp: str | None) -> list[dict]:
    """The checklist items of a document's full text, in order, with 1-based
    file lines. `timestamp` (the doc's) dates a done item that carries no
    `(done: YYYY-MM-DD)` suffix of its own; an open item is never dated."""
    doc_date = timestamp[:10] if timestamp else None
    items: list[dict] = []
    in_fence = False
    for number, line in enumerate(text.splitlines(), start=1):
        if _FENCE.match(line):
            in_fence = not in_fence
            continue
        if in_fence:
            continue
        match = _ITEM.match(line)
        if not match:
            continue
        status = "open" if match.group(1) == " " else "done"
        body = match.group(2).strip()
        done: str | None = None
        suffix = _DONE_SUFFIX.search(body)
        if suffix:
            body = body[: suffix.start()].rstrip()
            done = suffix.group(1)
        if status == "done" and done is None:
            done = doc_date
        items.append({"done": done if status == "done" else None, "line": number,
                      "status": status, "text": body})
    return items


def build_todos(docs: list[Document], root: str | Path) -> dict:
    """The t1/todos.json payload: every item of every to-do list, sorted by
    (path, line)."""
    root = Path(root)
    todos: list[dict] = []
    for doc in sorted(docs, key=lambda d: d.path):
        if not doc.todo:
            continue
        text = (root / doc.path).read_text(encoding="utf-8", errors="replace")
        for item in parse_todo_items(text, doc.timestamp):
            todos.append({"done": item["done"], "line": item["line"], "path": doc.path,
                          "status": item["status"], "text": item["text"]})
    return {"todos": todos}


def todo_counts(todos: list[dict], path: str | None = None) -> dict:
    """`{"open", "done"}` over the items — of one list when `path` is given."""
    scoped = [t for t in todos if path is None or t.get("path") == path]
    return {"open": sum(1 for t in scoped if t.get("status") == "open"),
            "done": sum(1 for t in scoped if t.get("status") == "done")}
